### PR TITLE
fix(netlify): pre-rendered 404 pages aren't shown

### PR DESCRIPTION
**Pull request by Skn0tt**
## Changes

Adds tests around 404 pages. Fixes prerendered 404 pages by embedding them into the SSR Handler and returning it when no route matches. There's other solutions to this, as we're discussing in https://github.com/withastro/adapters/issues/133, but this is a quick & easy fix to get 404 pages functioning.

## Testing

added two unit tests.

## Docs

It's a fix, so no docs needed. Related to https://github.com/withastro/adapters/issues/133.

### CODE REVIEW

File: packages/netlify/src/index.ts [hunk 1]
```diff
@@ -300,7 +301,11 @@ export default function netlifyIntegration(
 				logger.info('Emitted _redirects');
 
 				if (_config.output !== 'static') {
-					await writeSSRFunction();
+					let notFoundContent = undefined;
+					try {
+						notFoundContent = await readFile(new URL('./404.html', dir), 'utf8');
```
Comment (lilnasy) [inline packages/netlify/src/index.ts:306]: filesystem check would work reliably. That said, do you think we should prefer checking the routes object?

https://github.com/withastro/astro/pull/9648/files#diff-071ccfbdfd7b9bc803219179f88ae0bdb227a399bc4a481b7831c82b801d7a17R298

File: packages/netlify/src/ssr-function.ts [hunk 2]
```diff
@@ -8,15 +8,15 @@ applyPolyfills();
 // biome-ignore lint/complexity/noBannedTypes: safe to use in this case
 export type Args = {};
 
-const clientAddressSymbol = Symbol.for('astro.clientAddress');
-
 export const createExports = (manifest: SSRManifest, _args: Args) => {
 	const app = new App(manifest);
 
-	function createHandler(integrationConfig: { cacheOnDemandPages: boolean }) {
+	function createHandler(integrationConfig: { cacheOnDemandPages: boolean, notFoundContent?: string }) {
```
Comment (lilnasy) [inline packages/netlify/src/ssr-function.ts:14]: So it seems like the approach is to make the serverless function serve the static file by inlining 404.html into the code.

How feasible would it be to make the serverless function act only on certain routes?

This is how it's done in the vercel adapter: https://github.com/withastro/astro/pull/9648/files#diff-071ccfbdfd7b9bc803219179f88ae0bdb227a399bc4a481b7831c82b801d7a17R290-R296

Comment (lilnasy) [reply packages/netlify/src/ssr-function.ts:14]: This would address the (admittedly small) billing concern.

Comment (Skn0tt) [reply packages/netlify/src/ssr-function.ts:14]: That would require us to generate a `_redirects` file, which opens up some complexity that i'd prefer not to have. The single benefit from that, as I understand it, would be that there's no function execution for 404 pages. I'm unsure how valuable that really is, and if we do that change i'd prefer to do it in a separate PR in the future.

File: packages/netlify/src/index.ts (hunk 1 above)
Comment (lilnasy) [reply packages/netlify/src/index.ts:306]: I am happy as long as the users get an improvement, which this PR already provides. Just meant to point out that it is an option.

As for preferences, I am of the opinion it should be upto the more active contributor, which would be you.

Comment (alexanderniebuhr) [reply packages/netlify/src/index.ts:306]: I don't have any opinion here, fine with both approaches.

File: packages/netlify/src/ssr-function.ts [hunk 3]
```diff
@@ -27,13 +27,7 @@ export const createExports = (manifest: SSRManifest, _args: Args) => {
...
 			locals.netlify = { context };
 
-			const response = await app.render(request, { routeData, locals });
-
-			if (app.setCookieHeaders) {
-				for (const setCookieHeader of app.setCookieHeaders(response)) {
-					response.headers.append('Set-Cookie', setCookieHeader);
-				}
-			}
+			const response = await app.render(request, { routeData, locals, addCookieHeader: true, clientAddress: context.ip });
```
Comment (alexanderniebuhr) [inline packages/netlify/src/ssr-function.ts:30]: nit: This change seems unrelated to this PR. I would personally prefer to separate it.

**Review by alexanderniebuhr — COMMENTED** [review]
If I understand this correctly, only a `pre-rendered` 404 page will be embedded in the SSR function, correct?

If so, I don't get why we need this, my understanding from the discussion is that static assets are served before the SSR function, in that case a `pre-rendered` 404 page is a static asset (`html` file). Maybe someone could explain this again to me?

We would also like to get docs to look over the changeset, once we have the technical things aligned.

File: packages/netlify/src/ssr-function.ts (hunk 3 above)
Comment (Skn0tt) [reply packages/netlify/src/ssr-function.ts:30]: oof, that commit was still in my working copy and I didn't notice it! will remove it.

**Comment by Skn0tt** [comment]
> If so, I don't get why we need this, my understanding from the discussion is that static assets are served before the SSR function, in that case a pre-rendered 404 page is a static asset (html file). Maybe someone could explain this again to me?

Static assets are served before the SSR, that is true. That means if you go to `/404`, the 404 page will correctly be shown already without this PR. But if you go to any other non-existant page, like `/products/i-dont-exist`, there won't be a static asset on that route! So because the SSR function is configured as the fallback, it will be called, and now it needs to serve the 404 page. We could also serve a redirect to the 404 page, but that would be slightly different behaviour.

**Comment by alexanderniebuhr** [comment]
@Skn0tt  Ohhh, so Netlify does not fallback to `/404` if a `404.html` static asset exists in the deployment.

That's a different behaviour from Cloudflare, but then this PR makes sense to me.

In conclusion I guess Cloudflare has the reverse issue _(which we handle by the `_routes.json`)_, if you don't have a `404.html` but also don't route requests to the SSR function, Cloudflare will return an empty response. Which can't happen with Netlify, because the SSR function is the fallback.

**Comment by Skn0tt** [comment]
> Ohhh, so Netlify does not fallback to /404 if a 404.html static asset exists in the deployment. I would expect that..

Well, normally it does, but the Adapter explicitly sets the SSR function to serve all routes except for the ones with static assets:

https://github.com/withastro/adapters/blob/3854812c9bd3558857bb9084668925835dcf4ee0/packages/netlify/src/index.ts#L116

(see the `path: /*` and `preferStatic: true` options)

> Which can't happen with Netlify, because the SSR function is the fallback.

And that's exactly the reason why I prefer to do it the way around that I put here, even if it means that there's a function execution on every 404 :D

**Comment by alexanderniebuhr** [comment]
Got it. Thanks!
I already asked docs to get some feedback for the changeset.

File: .changeset/serious-kids-suffer.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@astrojs/netlify': patch
+---
+
+Fixes a bug in the Netlify Adapter where prerendered 404.astro pages weren't shown
```
Comment (sarah11918) [inline .changeset/serious-kids-suffer.md:5]: Would love for this to have a *tiny* bit more context, if it's a simple one liner...

> Fixes a bug in the Netlify Adapter where prerendered `404.astro` pages weren't displayed when/if...

If you can't answer that sentence in a few words, then it's fine! This feels like a bug that if people had it, they knew about it.  :smile: 

(If this didn't *always* happen, but the description of when it did is complicated or "it depends", you could say "where sometimes prerendered..." but not blocking.)

**Review by sarah11918 — COMMENTED** [review]
Docs review, as requested! 🫡

File: .changeset/serious-kids-suffer.md (hunk 4 above)
Comment (alexanderniebuhr) [reply .changeset/serious-kids-suffer.md:5]: @Skn0tt can you update the changeset?

Comment (Skn0tt) [reply .changeset/serious-kids-suffer.md:5]: I do! I'm sorry, got lost in a bunch of other work.

Comment (Skn0tt) [reply .changeset/serious-kids-suffer.md:5]:
```suggestion
Fixes a bug in the Netlify Adapter where prerendered 404.astro pages weren't shown on hybrid/server deployments.
```

File: packages/netlify/src/ssr-function.ts (hunk 3 above)
Comment (alexanderniebuhr) [reply packages/netlify/src/ssr-function.ts:30]: @Skn0tt do you still want to remove the commit? We can also keep it in the PR if you want.

Comment (Skn0tt) [reply packages/netlify/src/ssr-function.ts]: removed it!

File: .changeset/serious-kids-suffer.md
Comment (Skn0tt) [inline .changeset/serious-kids-suffer.md:1]: oh wow, this filename is quite dire 😨

**Comment by alexanderniebuhr** [comment]
@Skn0tt let me know when you're ready. I would merge it later in the evening :)

**Comment by Skn0tt** [comment]
Ready, feel free to merge!

**Review by alexanderniebuhr — APPROVED** [review]

**Comment by jamiemill** [comment]
My custom 404 page is showing raw html if I invoke it by going to a missing path like `/foo`, and is served as text/plain.

If I visit `/404` directly it's rendered as HTML.

Could this be related to this change?

**Comment by Skn0tt** [comment]
Uh-oh! Yes, that's related to this change. We're missing the content type here: https://github.com/withastro/adapters/blob/0a87e38eae63c043f053aea25a34bb81b6723895/packages/netlify/src/ssr-function.ts#L25

**Comment by Skn0tt** [comment]
here's the fix: https://github.com/withastro/adapters/pull/162
